File: lista_3/dh.py
import random
from typing import Callable

class FiniteFieldElement:
    ORDER = 1234577

    def __init__(self, value):
        self.value = value % self.ORDER

    def characteristic(self):
        return self.power_of_prime(self.ORDER)[0]

    def inv(self):
        t, newt, r, newr = 0, 1, self.ORDER, self.value
        while newr != 0:
            quotient = r // newr
            t, newt = newt, t - quotient * newt
            newr, r = r - quotient * newr, newr
        if r > 1:
            raise ArithmeticError(f"{self.value} is not invertible")
        if t < 0:
            t += self.ORDER
        return FiniteFieldElement(t)

    def __add__(self, other):
        return FiniteFieldElement((self.value + other.value) % self.ORDER)

    def __sub__(self, other):
        return FiniteFieldElement((self.value - other.value) % self.ORDER)

    def __mul__(self, other):
        return FiniteFieldElement((self.value * other.value) % self.ORDER)

    def __truediv__(self, other):
        return self * other.inv()

    def __str__(self):
        return str(self.value)

    def power_of_prime(self, n):
        for i in self.sieve_of_eratosthenes(n):
            if n % i == 0:
                count = 0
                while n % i == 0:
                    n //= i
                    count += 1
                if n == 1:
                    return i, count
                else:
                    raise ArithmeticError("Not a power of prime")
        raise ArithmeticError("Not a power of prime")

    def sieve_of_eratosthenes(self, n):
        is_prime = [True] * (n + 1)
        is_prime[0] = is_prime[1] = False
        p = 2
        while p * p <= n:
            if is_prime[p]:
                i = p * p
                while i <= n:
                    is_prime[i] = False
                    i += p
            p += 1
        return [i for i in range(n + 1) if is_prime[i]]

class DiffieHellmanSetup:
    CHARACTERISTIC = 1234577

    def __init__(self, constructor: Callable[[int], FiniteFieldElement]):
        self.constructor = constructor
        self.generator = self.get_generator()

    def get_generator(self):
        generator_value = random.randint(1, self.CHARACTERISTIC - 1)
        while not self.check(generator_value):
            generator_value = random.randint(1, self.CHARACTERISTIC - 1)
        return self.constructor(generator_value)

    def power(self, a: FiniteFieldElement, b: int):
        res = a
        while b > 0:
            if b % 2 == 1:
                res *= a
            a *= a
            b //= 2
        return res

    def check(self, suspect):
        i, tmp = 2, suspect
        while i * i <= tmp:
            if tmp % i == 0:
                if pow(suspect, (self.CHARACTERISTIC - 1) // i, self.CHARACTERISTIC) == 1:
                    return False
                tmp //= i
            else:
                i += 1
        if tmp > 1 and pow(suspect, (self.CHARACTERISTIC - 1) // tmp, self.CHARACTERISTIC) == 1:
            return False
        return True

class User:
    def __init__(self, name: str, dh_setup: DiffieHellmanSetup):
        self.name = name
        self.dh_setup = dh_setup
        self.secret = random.randint(0, 2**32)

    def get_public_key(self):
        return self.dh_setup.power(self.dh_setup.generator, self.secret)

    def set_key(self, a: FiniteFieldElement):
        self.key = self.dh_setup.power(a, self.secret)

    def encrypt(self, m: FiniteFieldElement):
        return m * self.key

    def decrypt(self, c: FiniteFieldElement):
        return c / self.key

# Utwórz obiekt DiffieHellmanSetup
dh = DiffieHellmanSetup(FiniteFieldElement)

# Utwórz użytkowników Alice i Bob
alice = User("Alice", dh)
bob = User("Bob", dh)

# Klucz publiczny Alica
alice_pub_key = alice.get_public_key()

# Klucz publiczny Boba
bob_pub_key = bob.get_public_key()

# Ustawienie klucza dla Alice i Bob
alice.set_key(bob_pub_key)
bob.set_key(alice_pub_key)

# Wspólny klucz
common_key = alice.key

# Wiadomość
message = FiniteFieldElement(2115)

# Zaszyfrowana wiadomość dla Alice
encrypted_message_alice = alice.encrypt(message)

# Odszyfrowana wiadomość dla Alice
decrypted_message_alice = alice.decrypt(encrypted_message_alice)

# Zaszyfrowana wiadomość dla Boba
encrypted_message_bob = bob.encrypt(message)

# Odszyfrowana wiadomość dla Boba
decrypted_message_bob = bob.decrypt(encrypted_message_bob)

# Wyświetlanie wyników
print("Perspektywa Alice:")
print("Generator:", dh.generator)
print("Klucz publiczny wysyłany:", alice_pub_key)
print("Klucz publiczny odbierany:", bob_pub_key)
print("Wspólny klucz obliczony przez Alice i Boba:", common_key)
print("Wiadomość:", message)
print("Zaszyfrowana wiadomość:", encrypted_message_alice)
print("Odszyfrowana wiadomość:", decrypted_message_alice)
print()

print("Perspektywa Boba:")
print("Generator:", dh.generator)
print("Klucz publiczny wysyłany:", bob_pub_key)
print("Klucz publiczny odbierany:", alice_pub_key)
print("Wspólny klucz obliczony przez Alice i Boba:", common_key)
print("Wiadomość:", message)
print("Zaszyfrowana wiadomość:", encrypted_message_bob)
print("Odszyfrowana wiadomość:", decrypted_message_bob)

File: lista_3/dh.java
import java.math.BigInteger;
import java.util.Random;
import java.util.function.Function;
import java.util.ArrayList;
import java.util.List;

public class dh {

    public static void main(String[] args) {
        FiniteFieldElement a = new FiniteFieldElement(19);
        FiniteFieldElement m = new FiniteFieldElement(2115);

        DiffieHellmanSetup<FiniteFieldElement> dh = new DiffieHellmanSetup<>(FiniteFieldElement::new);
        System.out.println("Generator: " + dh.getGenerator());

        // Utwórz dwóch użytkowników - Alice i Bob
        User<FiniteFieldElement> alice = new User<>(dh);
        User<FiniteFieldElement> bob = new User<>(dh);

        // Uzyskaj klucze publiczne dla Alice i Boba
        FiniteFieldElement alicePubKey = alice.getPublicKey();
        FiniteFieldElement bobPubKey = bob.getPublicKey();

        System.out.println("Alice Public Key: " + alicePubKey);
        System.out.println("Bob Public Key: " + bobPubKey);

        // Wymień klucze publiczne między Alice i Bobem
        alice.setReceiverPublicKey(bobPubKey);
        bob.setReceiverPublicKey(alicePubKey);

        // Zaszyfruj i odszyfruj wiadomość przez Alice i Boba
        System.out.println("Message: " + m);
        FiniteFieldElement c = alice.encrypt(m);
        System.out.println("Encrypted by Alice: " + c);
        FiniteFieldElement mDecrypted = bob.decrypt(c);
        System.out.println("Decrypted by Bob: " + mDecrypted);
    }

    private static final long CHARACTERISTIC = 1234577;

    public interface AlgebraicBody<T extends AlgebraicBody<T>> {
        T multiplyBy(T other);
        T divideBy(T other);
        T fromLong(long value);
    }

    public static boolean check(long suspect) {
        long i = 2;
        long tmp = suspect;

        while (i * i <= tmp) {
            if (tmp % i == 0) {
                if (BigInteger.valueOf(suspect).pow((int)((CHARACTERISTIC - 1) / i)).mod(BigInteger.valueOf(CHARACTERISTIC)).equals(BigInteger.ONE)) {
                    return false;
                }
                tmp /= i;
            } else {
                i += 1;
            }
        }
        if (tmp > 1 && BigInteger.valueOf(suspect).pow((int)((CHARACTERISTIC - 1) / tmp)).mod(BigInteger.valueOf(CHARACTERISTIC)).equals(BigInteger.ONE)) {
            return false;
        }
        return true;
    }

    public static class DiffieHellmanSetup<T extends AlgebraicBody<T>> {
        private T generator;

        public DiffieHellmanSetup(Function<Long, T> constructor) {
            Random rng = new Random();
            long generatorValue = rng.nextInt((int) CHARACTERISTIC - 1) + 1;
            while (!check(generatorValue)) {
                generatorValue = rng.nextInt((int) CHARACTERISTIC - 1) + 1;
            }
            this.generator = constructor.apply(generatorValue);
        }

        public T getGenerator() {
            return generator;
        }

        public T power(T a, long b) {
            T res = a;
            while (b > 0) {
                if (b % 2 == 1) {
                    res = res.multiplyBy(a);
                }
                a = a.multiplyBy(a);
                b /= 2;
            }
            return res;
        }
    }

    public static class User<T extends AlgebraicBody<T>> {
        private long secret;
        private DiffieHellmanSetup<T> dhsetup;
        private T key;
        private T receiverPublicKey;

        public User(DiffieHellmanSetup<T> dhsetup) {
            this.dhsetup = dhsetup;
            this.secret = new Random().nextLong();
            System.out.println("Secret: " + secret);
        }

        public T getPublicKey() {
            return dhsetup.power(dhsetup.getGenerator(), secret);
        }

        public void setReceiverPublicKey(T publicKey) {
            this.key = dhsetup.power(publicKey, secret);
            System.out.println("Key: " + key);
        }

        public T encrypt(T m) {
            return m.multiplyBy(key);
        }

        public T decrypt(T c) {
            return c.divideBy(key);
        }
    }

    public static class FiniteFieldElement implements AlgebraicBody<FiniteFieldElement> {
        private static final long ORDER = 1234577;
        private long value;

        public FiniteFieldElement(long value) {
            this.value = value % ORDER;
        }

        public long getValue() {
            return value;
        }

        public long characteristic() {
            long[] primePower = powerOfPrime(ORDER);
            return primePower[0];
        }

        private FiniteFieldElement inv() {
            long t = 0;
            long newt = 1;
            long r = ORDER;
            long newr = value;
            while (newr != 0) {
                long quotient = r / newr;
                long tmp = newt;
                newt = t - quotient * newt;
                t = tmp;
                tmp = newr;
                newr = r - quotient * newr;
                r = tmp;
            }
            if (r > 1) {
                throw new ArithmeticException(value + " is not invertible");
            }
            if (t < 0) {
                t += ORDER;
            }
            return new FiniteFieldElement(t);
        }

        public FiniteFieldElement add(FiniteFieldElement other) {
            return new FiniteFieldElement((value + other.value) % ORDER);
        }

        public FiniteFieldElement subtract(FiniteFieldElement other) {
            return new FiniteFieldElement((value + ORDER - other.value) % ORDER);
        }

        public FiniteFieldElement multiply(FiniteFieldElement other) {
            return new FiniteFieldElement((value * other.value) % ORDER);
        }

        public FiniteFieldElement divide(FiniteFieldElement other) {
            return multiply(other.inv());
        }

        private static long[] powerOfPrime(long n) {
            for (long i : sieveOfEratosthenes(n)) {
                if (n % i == 0) {
                    long count = 0;
                    while (n % i == 0) {
                        n /= i;
                        count++;
                    }
                    if (n == 1) {
                        return new long[]{i, count};
                    } else {
                        throw new ArithmeticException("Not a power of prime");
                    }
                }
            }
            throw new ArithmeticException("Not a power of prime");
        }

        private static List<Long> sieveOfEratosthenes(long n) {
            List<Boolean> isPrime = new ArrayList<>();
            for (int i = 0; i <= n; i++) {
                isPrime.add(true);
            }
            isPrime.set(0, false);
            isPrime.set(1, false);
            int p = 2;
            while (p * p <= n) {
                if (isPrime.get(p)) {
                    int i = p * p;
                    while (i <= n) {
                        isPrime.set(i, false);
                        i += p;
                    }
                }
                p++;
            }
            List<Long> primes = new ArrayList<>();
            for (int i = 0; i <= n; i++) {
                if (isPrime.get(i)) {
                    primes.add((long) i);
                }
            }
            return primes;
        }

        @Override
        public String toString() {
            return String.valueOf(value);
        }

        @Override
        public FiniteFieldElement multiplyBy(FiniteFieldElement other) {
            return multiply(other);
        }

        @Override
        public FiniteFieldElement divideBy(FiniteFieldElement other) {
            return divide(other);
        }

        @Override
        public FiniteFieldElement fromLong(long value) {
            return new FiniteFieldElement(value);
        }
    }
}
